Added TriangulatedMeshGenerator generator

roystgnr opened the pull request:
This refactors #19373 to use a poly2tri submodule instead of a poly2tri fork.

It passes tests for me, but I want to run it past CI here before I start squashing commits together to keep the fork out of the main git history.

Pinging @aaronoaks for another set of eyes, but the only actual change to his code in here should be trivial, that one range for loop where we now declare the element by reference, and unless someone sees other problems I just want to squash and merge later.

Closes #19372

Commit:
Added TriangulatedMeshGenerator meshgenerator interface to the poly2tri triangulation library.

closes #19372

Commit:
Removed development std::cout debug statements.

Commit:
Added input parameter checking for inner boundary id/name, verifying that the boundary exists on the mesh. Added regression test cases.

roystgnr committed:
Use libMesh contrib/poly2tri, not a forked copy

roystgnr committed:
Avoid copy in range for loop

This fixes a compiler warning for me, and might be better optimized.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/928429/) on 4f3f748 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20007/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20007/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a6076d579c4ea10b0c3bdb7d7561334faf24c241`

roystgnr committed:
clang-format style fixes

loganharbour commented on modules/reactor/doc/content/source/meshgenerators/TriangulatedMeshGenerator.md:
```diff
@@ -0,0 +1,33 @@
+
+## Overview
+
+The `TriangulatedMeshGenerator` was created to connect the functionality of the [poly2tri](https://github.com/jhasse/poly2tri) 2D triangulation library to the MOOSE mesh generation system. Since the motivation of this effort was to mesh the region between a reactor fuel core and the core periphery (typically a cylinder), it was designed to create a circular triangulation outer boundary and use th ...
```
I'm not a fan of the name. We need to reserve `TriangulatedMeshGenerator` for a general mesh generator in the framework.

roystgnr replied:
Strongly agree.  I'm not sure what to call this one, though.  TriangulatedAnnulusGenerator?  Except the region is only an annulus topologically, not geometrically.

eshemon replied:
We have another MG capability that is called PeripheralMeshGenerator, maybe this can be PeripheralTriangleMeshGenerator

loganharbour commented on modules/reactor/include/meshgenerators/TriangulatedMeshGenerator.h:
```diff
@@ -0,0 +1,64 @@
+   * @param inner_boundary_nodes vector of inner boundary nodes, input as empty list, populated by
+   * function.
+   */
+  void _create_sorted_boundary_node_list(std::unique_ptr<MeshBase> & mesh,
+                                         std::vector<Node *> & inner_boundary_nodes);
+
+  /**
+   * Deletes and clears a list of poly2tri Point pointers.
+   * @param point_list list of poly2tri Point pointers to be deleted.
+   */
+  void _clearPoints(std::vector<p2t::Point *> & point_list);
```
Needs a consistent naming scheme

roystgnr replied:
Agreed.  Moose convention is camelCaseWithInitialLowerCase for methods, snake_case for variables, single leading underscore for non-public:, correct?

loganharbour replied:
Yep, and we try to avoid `_` prefixes even for methods that are considered "internal"

roystgnr replied:
Ah, I think I get it - leading underscore for non-public members, no leading underscore for non-public methods, because we might want to make the methods public later but we'll never make the members public except via accessors?

loganharbour reviewed (commented):
Will do a more thorough review once requested; obviously still a lot of comments in here

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/928462/) on 8a278be wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20007/site)

This comment will be updated on new commits.

roystgnr committed:
Fix duplicate requirement line

roystgnr committed:
Workaround for execution with distributed meshes

I *think* my "we only ever care about distributing the mesh later" is
accurate here; if I'm wrong then eventually we'll need a fix rather than
just a workaround.

roystgnr committed:
Explicit libMesh includes

Hopefully this is enough to fix non-Unity builds

roystgnr committed:
Directly include CastUniquePtr header

By defining this overload of a std:: name outside the std:: namespace,
we should fix non-Unity builds while also making me die a little inside.

aaronoaks commented:
I was able to build with this fork and pass the reactor module tests, and the changes look good so far. What is the workflow moving forward for this? Is this PR meant to replace #19373 and we do the ongoing resolution (e.g., Logan's comments on name changes) here?

roystgnr commented:
> Is this PR meant to replace #19373 and we do the ongoing resolution (e.g., Logan's comments on name changes) here?

Assuming you're happy with that, yes.

aaronoaks commented:
Works for me. Should I close #19373 at this point? Also will you be lead on resolving the remaining changes, or will I need access to push to your fork?

roystgnr commented:
I will.  You probably don't have write access to my branch, but if you have any commits to add go ahead and push them to your repo and I'll cherry-pick.  And yeah, if you're happy with the main changes here, go ahead and close the old PR.

aaronoaks commented:
@eshemon @shikhar413 we will proceed with the integration with this PR from Roy and I will close the one I opened earlier (#19373).

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/928756/) on 3024bbe wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a6076d</th>
      <th colspan=4>#20007 3024bb</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.19%</td>
      <td>82.19%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72962</td>
      <td>72963</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15809</td>
      <td>15808</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20007/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20007/coverage/framework)


## Modules coverage

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>a6076d</th>
      <th colspan=4>#20007 3024bb</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.48%</td>
      <td>95.46%</td>
      <td>-0.02%</td>
      <td>95.12%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1712</td>
      <td>1829</td>
      <td>+117</td>
      <td>117</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>81</td>
      <td>87</td>
      <td>+6</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20007/coverage_diff/reactor.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20007/coverage/reactor)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20007/coverage/xfem)

</details>



This comment will be updated on new commits.

cticenhour reviewed (changes requested):
General review/reminder to add a note about this capability to the January newsletter now that `modules/doc/content/newsletter/2022_01.md` has made it to devel. Thanks!

roystgnr committed:
Rename Triangulated -> PeripheralTriangle

roystgnr committed:
Remove leading underscore

We use those for private member variables but not member functions

roystgnr committed:
PeripheralTriangleMeshGenerator newsletter blurb

eshemon commented on modules/doc/content/newsletter/2022_01.md:
```diff
@@ -7,6 +7,18 @@ for a complete description of all MOOSE changes.
+
+A new mesh generator available in the MOOSE reactors module, the
+[PeripheralTriangleMeshGenerator.md] allows users to take an arbitrary
+2-D input mesh (typically a slice of a reactor fuel core) and
```
```suggestion
2-D input mesh (typically a reactor core) and
```

roystgnr committed:
Update modules/doc/content/newsletter/2022_01.md

Co-authored-by: Emily Shemon <83663101+eshemon@users.noreply.github.com>

roystgnr committed:
Update modules/doc/content/newsletter/2022_01.md

Co-authored-by: Casey Icenhour <cticenho@ncsu.edu>

roystgnr committed:
Add example triangulated mesh to newsletter

I'm not sure whether a symlink or a relative path would be more
appropriate here, but this is a small png so a copy should be safe
enough.

aaronoaks commented:
@roystgnr @loganharbour Hey guys wanted to check in on this, let me know if there is anything you need from me to get the PR through.

loganharbour commented on modules/reactor/include/meshgenerators/PeripheralTriangleMeshGenerator.h:
```diff
@@ -0,0 +1,64 @@
+  const unsigned int _outer_circle_num_segments;
+  /// The boundary id for the generated mesh outer boundary.
+  const unsigned int _outer_boundary_id;
+  /// The boundary name of the generated mesh outer boundary.
+  // const std::string _outer_boundary_name;
```
```suggestion
```

loganharbour commented on modules/reactor/include/meshgenerators/PeripheralTriangleMeshGenerator.h:
```diff
@@ -0,0 +1,64 @@
+   * @param inner_boundary_nodes vector of inner boundary nodes, input as empty list, populated by
+   * function.
+   */
+  void createSortedBoundaryNodeList(std::unique_ptr<MeshBase> & mesh,
+                                    std::vector<Node *> & inner_boundary_nodes);
```
- Should be const
- No need to pass a `std::unique_ptr<MeshBase> &` here; just pass the `MeshBase &`
- Just return a vector instead of passing in the vector

loganharbour commented on modules/reactor/include/meshgenerators/PeripheralTriangleMeshGenerator.h:
```diff
@@ -0,0 +1,64 @@
+   * Deletes and clears a list of poly2tri Point pointers.
+   * @param point_list list of poly2tri Point pointers to be deleted.
+   */
+  void clearPoints(std::vector<p2t::Point *> & point_list);
```
Method should be const

loganharbour commented on modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:
```diff
@@ -0,0 +1,470 @@
+      "outer_boundary_id",
+      "outer_boundary_id>=0",
+      "The boundary id for the generated mesh outer boundary.");
+  // params.addParam<std::string>(
+  //   "outer_boundary_name",
+  //   "The boundary name of the generated mesh outer boundary.");
```
```suggestion
```

loganharbour commented on modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:
```diff
@@ -0,0 +1,470 @@
+    _outer_circle_num_segments(getParam<unsigned int>("outer_circle_num_segments")),
+    _outer_boundary_id(
+        isParamValid("outer_boundary_id") ? getParam<boundary_id_type>("outer_boundary_id") : 0),
+    // _outer_boundary_name(isParamValid("outer_boundary_name") ?
+    // getParam<std::string>("outer_boundary_name") : std::string("outside")),
```
```suggestion
```

loganharbour commented on modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:
```diff
@@ -0,0 +1,470 @@
+    _extra_circle_num_segments(getParam<std::vector<unsigned int>>("extra_circle_num_segments"))
+{
+  // confirm either id or name (exclusive or) for inner mesh boundary is provided
+  if ((!isParamValid("inner_boundary_id") && !isParamValid("inner_boundary_name")) ||
+      (isParamValid("inner_boundary_id") && isParamValid("inner_boundary_name")))
```
```suggestion
  if (isParamValid("inner_boundary_id") == isParamValid("inner_boundary_name"))
```

loganharbour commented on modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:
```diff
@@ -0,0 +1,470 @@
+  {
+    paramError(
+        "inner_boundary_id",
+        "Either 'inner_boundary_id' or 'inner_boundary_name' must be provided, but not both.");
```
```suggestion
        "Either 'inner_boundary_id' or 'inner_boundary_name' must be provided.");
```

roystgnr replied:
Isn't `but not both` accurate here?  If both are provided then we can't tell which to use.  (I suppose we could check the naming to see if the specifications were consistent and then carry on if so, but IMHO "just pick one or the other" is a reasonable user interface)

loganharbour replied:
Yeah that’s fine. That error comes out when none are provided too, but in that case extraneously saying “not both” is okay

loganharbour commented on modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:
```diff
@@ -0,0 +1,470 @@
+
+      // check if node has been created for point
+      // if not, create and add to map
+      if (point_node_map.find(point) == point_node_map.end())
+      {
+        point_node_map[point] = mesh->add_point(Point(point->x, point->y, 0.0));
+      }
```
```suggestion
      if (!point_node_map.count(point))
        point_node_map[point] = mesh->add_point(Point(point->x, point->y, 0.0));
```

loganharbour commented on modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:
```diff
@@ -0,0 +1,470 @@
+  {
+    boundary_info.add_node(point_node_map[point], _outer_boundary_id);
+  }
+  // boundary_info.sideset_name(_outer_boundary_id) = _outer_boundary_name;
+
+  // generate sidesets from nodesets
+  // boundary_info.build_side_list_from_node_list();
```
```suggestion
```

loganharbour reviewed (changes requested):
Minor comments. All looks good to me.

I would also prefer to use `std::unique_ptr<p2t::Point>` over `p2t::Point *` to remove the need to `delete`.

Sorry for the delay @aaronoaks

roystgnr commented:
> Sorry for the delay @aaronoaks

Likewise.

@loganharbour - other than my quibble above these look like good suggestions; I'll get them in now.

roystgnr commented:
Actually, that second helper function should be static, not just const, huh?  It doesn't use `this`, there won't likely be a need to change that in the future, and if it were to be changed the impact wouldn't be huge since it's not public.

roystgnr committed:
Better API for createSortedBoundaryNodeList

roystgnr committed:
clearPoints should be static

Not only is this even more conceptually correct than the original API or
the proposed const version, but in this version we recognize that we
don't even use `this` here.

roystgnr committed:
Update modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

roystgnr committed:
Remove unnecessary braces

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

roystgnr committed:
Cleaner logic

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

roystgnr committed:
Remove old comment

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

roystgnr committed:
Remove old comment

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

roystgnr committed:
Remove old comment

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/939944/) on 10d85ce wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20007/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20007/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ea4ed52a2966f3e58e60f4f105459fd7568f5d70`

roystgnr committed:
Fix clang format

roystgnr commented:
> I would also prefer to use std::unique_ptr<p2t::Point> over p2t::Point * to remove the need to delete.

Oh, sorry, I got to looking at later suggestions and forgot to respond to this one.

We definitely can't use `std::unique_ptr` directly here, because the point lists here are being passed directly to poly2tri APIs that take a vector of raw pointers.

We could be more Modern-C++-compliant, though: create `vector<p2t::Point>` objects to handle memory management, then just before we use the poly2tri APIs create `vector<p2t::Point *>` shims to them.  That'll probably be a little more efficient, too, and it'll still let us get rid of clearPoints.  Let me see how it looks.

roystgnr commented:
Oh, and I'll put the CDT on the stack while I'm at it.  It's always going to just be destructed at the end of that function regardless.

roystgnr committed:
Cleaner memory management for points

roystgnr committed:
Put CDT on the stack

This is faster and cleaner

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/940242/) on 981414d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20007/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20007/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 29221b23b92d59dce5bc02aa3da9d0fdef4fe350`

roystgnr committed:
clang-format patch

roystgnr committed:
Fix -Wpedantic warning

roystgnr commented:
Okay, there's no way anything in here broke Bison.  I think we're good.

@cticenhour - check out my newsletter comments?  If you're happy with those, then once we know Bison is squared away we can kick "Internal App tests" and merge right after.

cticenhour commented on modules/doc/content/newsletter/2022_01.md:
```diff
@@ -7,6 +7,25 @@ for a complete description of all MOOSE changes.
+
+[Figure 1](#triangle_steiner) shows an example triangulation around an existing core mesh.
+
+!media media/meshgenerators/triangle_abtr_steiner.png
```
```suggestion
!media modules/reactor/doc/content/media/meshgenerators/triangle_abtr_steiner.png
```
One comment on this - is there any way you can use the original copy of this file in the reactor module rather than duplicating the file in `modules/doc`? We build all modules when building the full website docs, sot it should be available.

roystgnr replied:
I'd love to, but if I interpret your question literally as a question then the answer is "I don't know".  I'm not familiar with this branch of markdown, and `find -name '*.md' | xargs grep '!media' | grep '\.\.'` came up blank for me so I was afraid using relative path names might not be possible.

cticenhour replied:
The way MooseDocs determines the available content is a bit weird. It needs to know that a requested file has been committed (it develops a file list using git), but before that it needs to know what locations are considered valid content. See [the link here](https://github.com/idaholab/moose/blob/ecbae1a77c12ff8dfe1af01e13f9b736857b7ba4/modules/doc/config.yml#L21) showing that we include the `modules/reactor/doc/content` folder in the main `modules/doc` documentation config for the website. Once MooseDocs knows what files are available, it copies what it needs to a website file structure so that everything is where the rendered webpages expect them to be (hence the truncated file path.....MooseDocs will look for files based on the "file roots" listed in the Content config). 

My suggestion above might actually be a poor one....You might only need to remove the `triangle_abtr_steiner.png` file you added in the `modules/doc/content/media` folder, and the build system will find the other one you already had added before. I should test this locally real quick before I ask you to make any changes (so as not to re-run every recipe before we need to).

cticenhour replied:
Removed my code suggestion. I confirmed that if you just remove the file `modules/reactor/doc/content/media/meshgenerators/triangle_abtr_steiner.png`, the docs will build successfully using the copy in `modules/reactor`.

cticenhour replied:
Just saw your new commit - everything looks good otherwise. Thanks @roystgnr!

cticenhour reviewed (changes requested):
Newsletter content looks good. One suggestion regarding the included media.

roystgnr committed:
Remove redundant image file

Our `make doc` is apparently smart enough to hunt down the correct file
even in its original position.

moosebuild commented:
Job [Internal App tests](https://civet.inl.gov/job/941667/) on e140ed2 : invalidated by @roystgnr

roystgnr commented on modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:
```diff
@@ -0,0 +1,453 @@
+    for (const int i : std::vector<int>{0, 1, 2})
+    {
+      // extract p2t::point from triangle
+      p2t::Point * point = triangle->GetPoint(i);
```
@aaronoaks -

Is poly2tri generating the ordering we expect here?  In my own poly2tri code I tried doing the same, set libMesh node i to p2t point i ... and it turns out I'm getting flipped elements, with clockwise instead of counter-clockwise node ordering.

It was actually quite tricky to run across that - libMesh generally assumes that your 2D elements might live in 3D space, makes sure their volume() isn't just half the cross product z component, likewise for JxW calculations, etc ... but even if it usually works fine I feel like this is going to bite us in some unexpected way (perhaps when we extrude the resulting mesh?) in the future.

roystgnr replied:
Actually, this might be a false alarm.  I am seeing a flipped element, but it may be from my own triangulation refinement code, not from poly2tri.  I'm adding some libMesh unit tests to make sure of that, though, and I still wouldn't mind independent confirmation of it.
